include/deal.II/distributed/cell_weights.h

```diff
@@ -38,80 +38,178 @@ namespace parallel
+   * object of this class needs to exist for every DoFHandler associated with
+   * the Triangulation we work on to achieve satisfying work balancing results.
+   * The following code snippet demonstrates how to achieve each cell being
+   * weighted by its current number of degrees of freedom. The connected
+   * weighting function may be changed anytime using the CellWeights::reinit()
+   * function.
    * @code
-   * parallel::CellWeights<dim, spacedim> cell_weights(hp_dof_handler);
-   * cell_weights.register_ndofs_weighting();
+   * parallel::CellWeights<dim, spacedim> cell_weights(
+   *   hp_dof_handler, parallel::CellWeights<dim, spacedim>::ndofs_weighting());
    * @endcode
-   * The weighting function can be changed anytime. Even more ambitious
-   * approaches are possible by submitting customized functions, e.g.
+   *
+   * On the other hand, you are also able to take care of handling the signal
+   * connection manually by using the static member function of this class. In
+   * this case, an analogous code example looks as follows.
    * @code
-   * cell_weights.register_custom_weighting(
-   *   [](const FiniteElement<dim, spacedim> &active_fe,
-   *      const typename hp::DoFHandler<dim, spacedim>::cell_iterator &)
-   *   -> unsigned int {
-   *   return 1000 * std::pow(active_fe.dofs_per_cell, 1.6);
-   * });
+   * boost::signals2::connection listener =
+   *   dof_handler.get_triangulation().signals.cell_weight.connect(
+   *     parallel::CellWeights<dim, spacedim>::make_weighting_callback(
+   *       dof_handler,
+   *       parallel::CellWeights<dim, spacedim>::ndofs_weighting());
    * @endcode
-   * The returned value has to be an unsigned integer and is thus limited by
-   * some large number. It is interpreted as the additional computational load
-   * of each cell. See Triangulation::Signals::cell_weight for a discussion on
-   * this topic.
+   *
+   * @note See Triangulation::Signals::cell_weight for more information.
    *
    * @note Be aware that this class connects the weight function to the
    * Triangulation during its construction. If the Triangulation
    * associated with the DoFHandler changes during the lifetime of the
    * latter, an assertion will be triggered in the weight_callback() function.
-   * It is recommended to create a separate object in this case and to destroy
-   * the previous one.
+   * Use CellWeights::reinit() to update dependencies.
    *
    * @ingroup distributed
-   * @author Marc Fehling, 2018
+   * @author Marc Fehling, 2018, 2019
    */
   template <int dim, int spacedim = dim>
   class CellWeights
   {
   public:
+    /**
+     * An alias that defines the characteristics of a function that can be used
+     * for weighting cells during load balancing.
+     */
+    using weighting_function = std::function<unsigned int(
+      const FiniteElement<dim, spacedim> &,
+      const typename hp::DoFHandler<dim, spacedim>::cell_iterator &)>;
```

Comment:
I suspect that the fe is to be used on the cell; so the fe is *subordinate* to the cell. This would suggest that one should reverse the order of arguments, unless that leads to an incompatible change.

Reply:
Agreed! The `FiniteElement` is the future finite element that will be associated to the cell after repartitioning.

In fact, we never use the `cell_iterator` in all of these weighting functions, but I wanted to leave it in case someone comes up with some fancy `weighting_function` on his own.